coil-core/src/commonMain/kotlin/coil3/Uri.kt

```diff
@@ -60,6 +103,52 @@ fun Uri(
     )
 }
 
+/**
+ * Create a new [Uri] by copying the properties of the original [Uri] and
+ * optionally modifying some of them.
+ */
+fun Uri.copy(
+    scheme: String? = this.scheme,
+    authority: String? = this.authority,
+    path: String? = this.path,
+    query: String? = this.query,
+    fragment: String? = this.fragment,
+): Uri {
+    return Uri(
+        scheme = scheme,
+        authority = authority,
+        path = path,
+        query = query,
+        fragment = fragment,
+        separator = this.separator,
+    )
+}
+
+/**
+ * Return a [UriBuilder] that is initialized with the components of this [Uri].
+ * This allows for convenient modification of the URI, such as adding query parameters.
+ *
+ * @return a new [UriBuilder] instance initialized with this URI's components.
+ */
+fun Uri.buildUpon(): UriBuilder {
```

Comment:
Let's make this a `Uri` class method instead of an extension function and rename it to `newBuilder` to be consistent with other classes in this repo.